### PR TITLE
add _created support: add createdTime() method to counter

### PR DESCRIPTION
**Pull request by andysim3d**
first step of issue #685 

Signed-off-by: Pengfei Zhang <zhangpengfei@google.com>

### CODE REVIEW

**Comment by andysim3d** [comment]
@beorn7 Please take a look since you are the original owner of that issue.

**Comment by beorn7** [comment]
While I opened this issue, this has to be figured out between the OpenMetrics experts and the current maintainers of prometheus/client_golang.

From a very general perspective, I don't think `Created()` should be a method (similar to not making the counter value itself accessible via some kind of `Get()` method).

File: prometheus/counter.go [hunk 1]
```diff
@@ -41,6 +41,9 @@ type Counter interface {
 	// Add adds the given value to the counter. It panics if the value is <
 	// 0.
 	Add(float64)
+
+	// Get _created suffixed metric. Records when it got first exposed.
+	Created() float64
```
[the remark this answers is not in the document]
Comment (beorn7) [reply prometheus/counter.go:46]: Indeed. I guess "when it got initialized" might be a more accurate description.

**Review by kakkoyun — COMMENTED** [review]
We haven't discussed OpenMetrics support among maintainers yet. cc @bwplotka 
I'll add this to the milestone. And ask for your further patience for the decision.

**Comment by pintohutch** [comment]
Hello,

Is this still being discussed?

**Comment by logicalhan** [comment]
Can we please do this? `_created` timeseries is a necessary prerequisite for upstream Kubernetes to even consider supporting prometheus' open-metrics format.

File: prometheus/counter.go [hunk 2]
```diff
@@ -88,8 +89,9 @@ type counter struct {
 	// valInt stores values that are exact integers. Both have to go first
 	// in the struct to guarantee alignment for atomic operations.
 	// http://golang.org/pkg/sync/atomic/#pkg-note-BUG
-	valBits uint64
-	valInt  uint64
+	valBits     uint64
+	valInt      uint64
+	createdTime uint64
```
Comment (bwplotka) [inline prometheus/counter.go:94]: I don't think this field is ever used, no?

**Review by bwplotka — COMMENTED** [review]
Hi, 

Thanks for this. We decided on Prometheus DevSummit to pursue proto version of OpenMetrics format as the main one. So I would implement that first.

For OM v1.0 text format, there are many weird conflicts and incompatibilities (e.g spam of extra metric on all Prometheus versions), so we can't really support it in client_golang and definitely not by default. I would suggest waiting on text improvement for OM e.g in OM 2.0 and rely on proto only. 

WDYT?

File: prometheus/counter.go (hunk 2 above)
Comment (pintohutch) [reply prometheus/counter.go:94]: It's initialized on L166 in `initCreated()` call.

Presumably, a forthcoming change would use it for serving in a Prometheus counter metric.

Comment (bwplotka) [reply prometheus/counter.go:94]: Yup, but we would need this to be either in this PR ideally so we see how end solution would look like 🤗 

Definitely it's good direction!

**Comment by bwplotka** [comment]
BTW, @ArthurSens, our mentee works on adding created timestamp with tests https://github.com/prometheus/client_golang/pull/1044

**Comment by bwplotka** [comment]
Done in https://github.com/prometheus/client_golang/pull/1044, so closing this PR. Thanks anyway!

**Comment by logicalhan** [comment]
> Done in #1044, so closing this PR. Thanks anyway!

this is PR #0144...

**Comment by ArthurSens** [comment]
I believe he meant #1313

**Pull request closed without merging**